builder/qemu/step_run.go

```diff
@@ -66,140 +70,187 @@ func (s *stepRun) Cleanup(state multistep.StateBag) {
 	}
 
-	rawVersion, err := driver.Version()
-	if err != nil {
-		return nil, err
+	// Configure "-vnc" arguments
+	// vncPort is always set in stepConfigureVNC, so we don't need to
+	// defensively assert
+	vncPort := state.Get("vnc_port").(int)
+	vncIP := config.VNCBindAddress
+
+	vncPort = vncPort - config.VNCPortMin
+	vnc := fmt.Sprintf("%s:%d", vncIP, vncPort)
+	if config.VNCUsePassword {
+		vnc = fmt.Sprintf("%s:%d,password", vncIP, vncPort)
 	}
-	qemuVersion, err := version.NewVersion(rawVersion)
-	if err != nil {
-		return nil, err
+	defaultArgs["-vnc"] = vnc
+
+	// Track the connection for the user
+	vncPass, _ := state.Get("vnc_password").(string)
+
+	message = getVncConnectionMessage(config.Headless, vnc, vncPass)
+	if message != "" {
+		s.ui.Message(message)
 	}
-	v2 := version.Must(version.NewVersion("2.0"))
 
-	if qemuVersion.GreaterThanOrEqual(v2) {
-		if config.DiskInterface == "virtio-scsi" {
-			if config.DiskImage {
-				deviceArgs = append(deviceArgs, "virtio-scsi-pci,id=scsi0", "scsi-hd,bus=scsi0.0,drive=drive0")
-				driveArgumentString := fmt.Sprintf("if=none,file=%s,id=drive0,cache=%s,discard=%s,format=%s", imgPath, config.DiskCache, config.DiskDiscard, config.Format)
-				if config.DetectZeroes != "off" {
-					driveArgumentString = fmt.Sprintf("%s,detect-zeroes=%s", driveArgumentString, config.DetectZeroes)
-				}
-				driveArgs = append(driveArgs, driveArgumentString)
-			} else {
-				deviceArgs = append(deviceArgs, "virtio-scsi-pci,id=scsi0")
-				diskFullPaths := state.Get("qemu_disk_paths").([]string)
-				for i, diskFullPath := range diskFullPaths {
-					deviceArgs = append(deviceArgs, fmt.Sprintf("scsi-hd,bus=scsi0.0,drive=drive%d", i))
-					driveArgumentString := fmt.Sprintf("if=none,file=%s,id=drive%d,cache=%s,discard=%s,format=%s", diskFullPath, i, config.DiskCache, config.DiskDiscard, config.Format)
-					if config.DetectZeroes != "off" {
-						driveArgumentString = fmt.Sprintf("%s,detect-zeroes=%s", driveArgumentString, config.DetectZeroes)
-					}
-					driveArgs = append(driveArgs, driveArgumentString)
-				}
+	// Configure "-m" memory argument
+	defaultArgs["-m"] = fmt.Sprintf("%dM", config.MemorySize)
+
+	// Configure "-smp" processor hardware arguments
+	if config.CpuCount > 1 {
+		defaultArgs["-smp"] = fmt.Sprintf("cpus=%d,sockets=%d", config.CpuCount, config.CpuCount)
+	}
+
+	// Configure "-fda" floppy disk attachment
+	if floppyPathRaw, ok := state.GetOk("floppy_path"); ok {
+		defaultArgs["-fda"] = floppyPathRaw.(string)
+	} else {
+		log.Println("Qemu Builder has no floppy files, not attaching a floppy.")
+	}
+
+	// Configure GUI display
+	if !config.Headless {
+		if s.atLeastVersion2 {
+			// FIXME: "none" is a valid display option in qemu but we have
+			// departed from the qemu usage here to instaed mean "let qemu
+			// set a reasonable default". We need to deprecate this behavior
+			// and let users just set "UseDefaultDisplay" if they want to let
+			// qemu do its thing.
```

Comment:
This is an error I found while working on this code and doing some research -- fixing it will break BC so is outside the scope of this PR.